### ISSUE TITLE
PROCESS CHANGE: Adopting lazy consensus via Github PRs

### ISSUE THREAD

**Issue by julz** [knative/community#478]
# Background

_Note: This is a modification of the Proposal in https://github.com/knative/community/issues/474. The vast bulk remains the same (it's an excellent proposal), but this issue modifies it slightly to propose a model based on Kubernetes-style markdown-based "KNEPs", primarily discussed in github and with the repo as the source of truth, rather than a focus on mailing list discussion._

Transparent decision making and effective communication are key for the success of any FOSS. While the project contributors, approvers and WG leads do their best to ensure we have a healthy and inclusive community, there's still big room for improvement. This is a proposal that came after a long discussion on the Knative Developers Mailing List, and on https://github.com/knative/community/issues/474. 

**This proposal is modified to replace the requirement to "use Knative Developers mailing list as the "formal" communication channel for decisions, their technical discussions, and their status updates (approve, reject, etc)" with a similar requirement based on using the repo as the source of truth.**

# Proposal

## Changes

- Adopting lazy consensus as a decision-making process. To enable lazy consensus to operate well, proposals should be notified to the mailing list and slack as well as the corresponding issue (which is the source of truth).
- Using Feature Track docs or "KNEPs" _(PR-ed as markdown to a knative/enhancements repo)_. The repo is the source of truth.
- Using Github issues and PRs as the "formal" communication channel for decisions, their technical discussions, and their status updates (approve, reject, etc). Slack, mailing lists and meetings are fine for socialising issues before proposal etc (and required to notify to achieve lazy consensus), but formal discussion and decision making should happen asynchronously on an issue, and context from other locations should be reflected on the repo ("or it didn't happen")

**Minimum time to revisit : 6 months**

Nothing is set in stone, but we think this is the minimum time to try out the proposed changes and give them a fair shot before revisiting them.

## Benefits

1. Clear expectation of what is the single source of truth of the status. Is it a GitHub issue that links to the FT doc? the FT doc itself? (i.e copy key comments from the GitHub issues into the doc?) else? (Markdown with the FT akin to TEP)? ML? --> **It is the merged KNEP in the repo.**
1. Agreeing on the minimal decisions communication expectations and narrowing them to a single communication channel that is:
     - Fully asynchronous, ideally a contributor (not necessarily WG leads) wouldn't even need to join any meeting to smoothly propose a FT, follow up its approval/adoption and implement it.
     - Preserves history.
     - Have a high SNR.
     - Have a high delivery guarantee. Email has a "delivery guarantee", "you've been informed". That's important for adopting "lazy consensus", cause literally every single contributor gets his own copy of proposals/decisions. **To achieve this we will require that for lazy consensus both the mailing list and relevant slack channels are notified**.
     - Discoverable/linkable/searchable for the intended users, i.e. "contributors"
     - --> **A KNEP-based process based around Github issues and PRs gives us all of these things, while also being easily accessible and discoverable to new contributors. The repo is the source of truth**.
1. Enhancing decision making by adopting a time tested, transparent decision-making process used by well established entities like ASF and Kuberenetes itself. -> **This is loosely modelled on the Kubernetes KEP + lazy consensus processes, which seem to have worked ~reasonably well**.
1. Promoting more inclusion for users across different time zones.
1. Promoting community trust because in lazy consensus a contributor just assumes they have the community’s support unless someone says otherwise.
1. **Using a knative/enhancements repo for major enhancements allows subscription to one repo for major enhancements across the project, which is important to eventing which has many repos. Using PR-ed markdown for proposals allows line-by-line comments and discussion, as well as an in-repo, discoverable, linkable record of the status and context (and avoids the pain-points around google docs and drive for this use case)**.

## Problems

These are well-identified and discussed in https://github.com/knative/community/issues/474.

## Adoption steps

It's understood that changing people and their behavior is hard and takes time. These are the steps needed to start introducing those changes

1. Create a one pager about lazy consensus in Knative and link it to all key repos documents.
1. Create a knative/enhancements repo to store KNEPs. This replaces the google drive based approach. 
   - BONUS: people don't have to sign up to a google group to view these docs. 
   - BONUS 2: we don't have to use google docs comment functionality, which is confusing and loses context.
1. WG leads should start encouraging that technical discussions take place over GitHub issues/PRs and refrain from taking decisions in meetings or on slack. Note: it's ok to have discussions on slack or lists, and we expect most ideas to be socialised there before becoming a formal proposal, but all important context, and all decisions, should be reflected back to the repo. "If it's not on the repo, it didn't happen".

### Why not mailing lists?

The advantages of mailing lists (in the context of discussing and documenting technical decisions and proposals) for those who want them can be achieved by subscribing to email notifications from GitHub. OTOH using the repo as the source of truth and context provides a number of advantages not present for email. Github issues are easily searched by new contributors (who may not even discover a mailing list at first), can be selectively watched, and have a rich notification system. Newer contributors are likely to be familiar with GitHub, but are much less likely to have google groups set up and to be comfortable applying appropriate email filters etc. This is especially the case where a contributor wants to participate or gain context on only one or two issues. Discoverability of past context is very high with GitHub issues, and they are easy to cross-reference to eventual PRs. All context and state is in one place: the repo, and a single search turns up all context (small issues and "big") on a topic. Keeping the bar to participation low, and easing the process of gaining context on past decisions, is important to the project.

### Why not google doc-based feature docs?

The comment system on google docs loses context and is hard to follow (especially when things are "resolved"). It's possible to find old comments but the UI makes it a struggle. It also splits discussions across yet another medium and makes people sign up to a google group to get access to things. Having all context be in the repo centralises discussion, clarifies current status, and makes it easy to get context on a past proposal or decision.

### What about slack/mailing lists/meeting etc

Slack, mailing lists and meetings continue and are both fine and good for e.g. socialising an idea, getting some initial feedback etc, but significant discussions - and especially decisions - should be in the repo, and any context from them should be reflected on an issue/PR ("or it didn't happen" 🙂).

**Comment by devguyio** [comment]
I think we should close this for #579

**Comment by julz** [comment]
agreed @devguyio 👍 

/close
